src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -3,138 +3,700 @@
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out bool, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out bool value)
+        public static unsafe int TryParse(ref this BufferReader<byte> reader, out bool value, char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
 
             // For other types (int, etc) we won't know if we've consumed all of the type
             // ("235612" can be split over segments, for example). For bool, Utf8Parser
             // doesn't care what follows "True" or "False" and neither should we.
-            if (Utf8Parser.TryParse(unread, out value, out int consumed))
+            if (Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                reader.Advance(consumed);
+                return consumed;
+            }
+            else if (unread.Length >= MaxBoolLength)
+            {
+                return 0;
+            }
+
+            return TryParseSlow(ref reader, out value, standardFormat);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private unsafe static int TryParseSlow(ref BufferReader<byte> reader, out bool value, char standardFormat)
+        {
+            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+
+            byte* buffer = stackalloc byte[MaxBoolLength];
+            Span<byte> tempSpan = new Span<byte>(buffer, MaxBoolLength);
+            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed, standardFormat))
+            {
+                reader.Advance(consumed);
+            }
+
+            return consumed;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static unsafe int TryParseSlow<T>(
+            ref BufferReader<byte> reader,
+            out T value,
+            char standardFormat,
+            int bufferSize,
+            ParseDelegate<T> parseDelegate) where T: unmanaged
+        {
+            // Note that in calling this with the expected max size of a normal instance of the given
+            // type may do more work than needed for the degenerate case of zero prefaced numbers at
+            // the end of the span (e.g. 00000000000000000000000000123). Doing the extra check slows
+            // down the perf of "normal" numbers so we just let the "bad" numbers eat up more cycles.
+
+            // stackalloc is separated as we can't represent in C# that Peek will not capture the span.
+            byte* stackBytes = stackalloc byte[MaxParseBytes];
```

Comment:
Does it make sense to separate out the common path (where we stackalloc exactly bufferSize rather than MaxParseBytes)? If we fail on our first attempt, we can then fall back to the doubling loop below.

Reply:
No, because the logic would be really complicated to avoid stack pressure. Every `stackalloc` adds more to the stack, even in a loop.